Enabling user in contentChangeDetails

shtukas opened the pull request:
This change brings the setting of the user in the contentChangeDetails of atoms.
It builds on the work made in the previous change :)

shtukas committed:
Enabling setting user of atom contentChangeDetails

    This change brings the settting of the user in the contentChangeDetails of atoms.
    It builds on the work made in the previous change :)

philmcmahon commented on explainer-server/app/models/ExplainerStore.scala:
```diff
@@ -18,17 +18,19 @@ class ExplainerStore @Inject() (config: Config) extends ExplainerAtomImplicits
   val explainerDB = new ExplainerDB(config)
 
-  def buildAtomWithDefaults(id: String, explainerAtom: ExplainerAtom, revision: Long): Atom = {
+  def buildAtomWithDefaults(user: com.gu.pandomainauth.model.User, id: String, explainerAtom: ExplainerAtom, revision: Long): Atom = {
```
import the class to clean this up and maybe move to end of parameter list

philmcmahon replied:
take contentChangeDetails as parameter

shtukas replied:
Both done!

philmcmahon commented on explainer-server/app/models/ExplainerStore.scala:
```diff
@@ -46,16 +48,16 @@ class ExplainerStore @Inject() (config: Config) extends ExplainerAtomImplicits
-    val explainer = buildAtomWithDefaults(uuid, explainerAtom, 1)
+    val explainer = buildAtomWithDefaults(user, uuid, explainerAtom, 1)
     explainerDB.store(explainer)
     Future(explainer)
```
Add a publish function here which updates the published field of contentChangeDetails

shtukas committed:
Updates:

- import the class and move to end of parameter list.
- take contentChangeDetails as parameter.
- Add a publish function here which updates the published field of contentChangeDetails.

shtukas committed:
code reformatting

philmcmahon commented on explainer-server/app/models/ExplainerStore.scala:
```diff
@@ -46,18 +48,37 @@ class ExplainerStore @Inject() (config: Config) extends ExplainerAtomImplicits
     val uuid = java.util.UUID.randomUUID.toString
     val explainerAtom = ExplainerAtom("-", "-", DisplayType.Expandable)
-    val explainer = buildAtomWithDefaults(uuid, explainerAtom, 1)
+    val contentChangeDetails = ContentChangeDetails(
```
Add created ChangeRecord here

shtukas committed:
setting all the fields of ContentChangeDetails

shtukas committed:
pandaUserToAtomUser

shtukas committed:
Code refactoring: introducing contentChangeDetailsBuilder

... Phil made me do it :)

philmcmahon commented on explainer-server/app/models/ExplainerStore.scala:
```diff
@@ -14,21 +14,41 @@ import scala.concurrent.Future
+      created      = buildChangeRecord(existingContentChangeDetails.flatMap(_.created)     , updateCreated),
+      lastModified = buildChangeRecord(existingContentChangeDetails.flatMap(_.lastModified), updateLastModified),
+      published    = buildChangeRecord(existingContentChangeDetails.flatMap(_.published)   , updatePublished),
+      revision     = existingContentChangeDetails.map(_.revision).getOrElse(1)
```
when updating, we need to update this. maybe 

`existingContentChangeDetails.map(_.revision).getOrElse(0)+1`

shtukas committed:
correcting revision number computation

philmcmahon commented on explainer-server/app/models/ExplainerStore.scala:
```diff
@@ -46,18 +66,31 @@ class ExplainerStore @Inject() (config: Config) extends ExplainerAtomImplicits
+    explainerDB.load(id).map{ explainer =>
+      val newExplainerAtom = ExplainerAtom(explainer.tdata.title, explainer.tdata.body, explainer.tdata.displayType)
+      val contentChangeDetails = contentChangeDetailsBuilder(user, explainer.contentChangeDetails, false, false, true)
+      val updatedExplainer = buildAtomWithDefaults(explainer.id, newExplainerAtom, contentChangeDetails, user)
```
newExplainerAtom is unnecessary here, you can just user explainer.data

shtukas committed:
Correcting typo

philmcmahon commented on explainer-server/app/models/ExplainerStore.scala:
```diff
@@ -46,18 +66,31 @@ class ExplainerStore @Inject() (config: Config) extends ExplainerAtomImplicits
+
+  def publish(id: String, user: PandaUser): Future[Atom] = {
+    explainerDB.load(id).map{ explainer =>
+      val newExplainerAtom = ExplainerAtom(explainer.tdata.title, explainer.tdata.body, explainer.tdata.displayType)
```
this shouldn't be necessary - just use explainer.data

shtukas committed:
reformatting

philmcmahon commented:
👍
